lambda/transl_list_comprehension.ml

```diff
@@ -0,0 +1,302 @@
+  ; add_bindings : lambda -> lambda
+  (** Any extra bindings that should be present in the body of this iterator,
+      for use by nested pieces of the translation; used if the user specifies a
+      complex pattern in a [for]-[in] iterator. *)
+  }
+
+(** Translates an iterator ([Typedtree.comprehension_iterator]), one piece of a
+    [for ... and ... and ...] expression, into Lambda.  This translation is into
+    a [translated_iterator], not just a Lambda term, because the iterator
+    desugars into a higher-order function which is applied to another function
+    containing the body of the iteration; that body function can't be filled in
+    until the rest of the translations have been done. *)
+let iterator ~transl_exp ~scopes = function
+  | Texp_comp_range { ident; pattern = _; start; stop; direction } ->
+      (* We have to let-bind [start] and [stop] so that they're evaluated in the
+         right (i.e., left-to-right) order *)
+      let transl_bound var bound =
+        Let_binding.make_id'
+          (Immutable Strict) Pintval
+          var (transl_exp ~scopes bound)
+      in
+      let start = transl_bound "start" start in
+      let stop  = transl_bound "stop"  stop  in
+      { builder      = (match direction with
+                        | Upto   -> rev_dlist_concat_iterate_up
+                        | Downto -> rev_dlist_concat_iterate_down)
+      ; arg_lets     = [start; stop]
+      ; element      = ident
+      ; element_kind = Pintval
+      ; add_bindings = Fun.id
+      }
+  | Texp_comp_in { pattern; sequence } ->
+      let iter_list =
+        Let_binding.make_id' (Immutable Strict) Pgenval
+          "iter_list" (transl_exp ~scopes sequence)
+      in
+      (* Create a fresh variable to use as the function argument *)
+      let element = Ident.create_local "element" in
+      { builder      = rev_dlist_concat_map
+      ; arg_lets     = [iter_list]
+      ; element
+      ; element_kind = Typeopt.value_kind pattern.pat_env pattern.pat_type
+      ; add_bindings =
+          (* CR aspectorzabusky: This has to be at [value_kind] [Pgenval],
+             right, since we don't know more specifically? *)
+          Matching.for_let
+            ~scopes pattern.pat_loc (Lvar element) pattern Pgenval
+      }
+
+(** Translates a list comprehension binding
+    ([Typedtree.comprehension_clause_binding]) into Lambda.  At parse time,
+    iterators don't include patterns and bindings do; however, in the typedtree
+    representation, the patterns have been moved into the iterators (so that
+    range iterators can just have an [Ident.t], for translation into for loops),
+    so bindings are just like iterators with a possible annotation.  As a
+    result, this function is essentially the same as [iterator], which see. *)
+let binding ~transl_exp ~scopes { comp_cb_iterator; comp_cb_attributes = _ } =
+  (* CR aspectorzabusky: What do we do with attributes here? *)
```

Comment:
The old code for type-checking for loop indices drops attributes -- they're never looked at. So it's reasonable to do the same for range iterators. But `in` should preserve them. It looks like `Translattribute.add_function_attributes` is a good starting place.